src/thread_task.rs

```diff
@@ -626,23 +622,25 @@ unsafe fn delayed_fg_task<'l, 'ttd: 'l>(
             if ttd.cond_signaled.load(Ordering::SeqCst) != 0 {
                 ttd.cond.notify_one();
             }
-            let mut delayed_fg = ttd.delayed_fg.try_write().unwrap();
-            match (*out).p.bpc {
+            let mut delayed_fg_guard = ttd.delayed_fg.try_write().unwrap();
+            // re-borrow to allow independent field borrows
+            let delayed_fg = &mut *delayed_fg_guard;
+            match delayed_fg.out.p.bpc {
                 #[cfg(feature = "bitdepth_8")]
                 bpc @ 8 => {
                     rav1d_prep_grain::<BitDepth8>(
                         &Rav1dDSPContext::get(bpc).as_ref().unwrap().fg,
-                        &mut *out,
-                        &*in_0,
+                        &delayed_fg.out,
+                        &delayed_fg.in_0,
                         BitDepth8::select_mut(&mut delayed_fg.grain),
                     );
                 }
                 #[cfg(feature = "bitdepth_16")]
                 bpc @ 10 | bpc @ 12 => {
                     rav1d_prep_grain::<BitDepth16>(
                         &Rav1dDSPContext::get(bpc).as_ref().unwrap().fg,
-                        &mut *out,
-                        &*in_0,
+                        &mut delayed_fg.out,
```

Comment:
I think only `rav1d_apply_grain_row` needs interior `DisjointMut`ability, since disjoint rows are accessed concurrently, but for `rav1d_prep_grain`, we should be able to pass a `&mut out` directly.

Reply:
Sure, I debated removing that mutability, because it has no effect currently and is not enforced (because the picture data are raw pointers).